File: LogisimPlus/LogisimPlus/logisim/circuit/appear/CircuitPins.cs
// ====================================================================================================
// Produced by the Free Edition of Java to C# Converter.
// To produce customized conversions, purchase a Premium Edition license:
// https://www.tangiblesoftwaresolutions.com/product-details/java-to-csharp-converter.html
// ====================================================================================================

using System.Collections.Generic;

/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

namespace logisim.circuit.appear
{

	using ReplacementMap = logisim.circuit.ReplacementMap;
	using Component = logisim.comp.Component;
	using ComponentEvent = logisim.comp.ComponentEvent;
	using ComponentListener = logisim.comp.ComponentListener;
	using logisim.data;
	using AttributeEvent = logisim.data.AttributeEvent;
	using AttributeListener = logisim.data.AttributeListener;
	using Instance = logisim.instance.Instance;
	using StdAttr = logisim.instance.StdAttr;
	using Pin = logisim.std.wiring.Pin;

	public class CircuitPins
	{
		private class MyComponentListener : ComponentListener, AttributeListener
		{
			private readonly CircuitPins outerInstance;

			public MyComponentListener(CircuitPins outerInstance)
			{
				this.outerInstance = outerInstance;
			}

			public virtual void endChanged(ComponentEvent e)
			{
				outerInstance.appearanceManager.updatePorts();
			}

			public virtual void componentInvalidated(ComponentEvent e)
			{
			}

			public virtual void attributeListChanged(AttributeEvent e)
			{
			}

			public virtual void attributeValueChanged(AttributeEvent e)
			{
// JAVA TO C# CONVERTER WARNING: Java wildcard generics have no direct equivalent in C#:
// ORIGINAL LINE: logisim.data.Attribute<?> attr = e.getAttribute();
				Attribute attr = e.Attribute;
				if (attr == StdAttr.FACING || attr == StdAttr.LABEL || attr == Pin.ATTR_TYPE)
				{
					outerInstance.appearanceManager.updatePorts();
				}
			}
		}

		private PortManager appearanceManager;
		private MyComponentListener myComponentListener;
		private HashSet<Instance> pins;

		internal CircuitPins(PortManager appearanceManager)
		{
			this.appearanceManager = appearanceManager;
			myComponentListener = new MyComponentListener(this);
			pins = new HashSet<Instance>();
		}

		public virtual void transactionCompleted(ReplacementMap repl)
		{
			// determine the changes
			HashSet<Instance> adds = new HashSet<Instance>();
			HashSet<Instance> removes = new HashSet<Instance>();
			Dictionary<Instance, Instance> replaces = new Dictionary<Instance, Instance>();
			foreach (Component comp in repl.Additions)
			{
				if (comp.Factory is Pin)
				{
					Instance @in = Instance.getInstanceFor(comp);
					bool added = pins.Add(@in);
					if (added)
					{
						comp.addComponentListener(myComponentListener);
						@in.AttributeSet.addAttributeListener(myComponentListener);
						adds.Add(@in);
					}
				}
			}
			foreach (Component comp in repl.Removals)
			{
				if (comp.Factory is Pin)
				{
					Instance @in = Instance.getInstanceFor(comp);
					bool removed = pins.remove(@in);
					if (removed)
					{
						comp.removeComponentListener(myComponentListener);
						@in.AttributeSet.removeAttributeListener(myComponentListener);
						ICollection<Component> rs = repl.getComponentsReplacing(comp);
						if (rs.Count == 0)
						{
							removes.Add(@in);
						}
						else
						{
							Component r = rs.GetEnumerator().next();
							Instance rin = Instance.getInstanceFor(r);
							adds.remove(rin);
							replaces[@in] = rin;
						}
					}
				}
			}

			appearanceManager.updatePorts(adds, removes, replaces, Pins);
		}

		public virtual ICollection<Instance> Pins
		{
			get
			{
				return new List<Instance>(pins);
			}
		}
	}

}

File: logisim/app/src/main/java/logisim/circuit/appear/CircuitPins.java
/* Copyright (c) 2010, Carl Burch. License information is located in the
 * logisim.Main source code and at www.cburch.com/logisim/. */

package logisim.circuit.appear;

import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Map;
import java.util.Set;

import logisim.circuit.ReplacementMap;
import logisim.comp.Component;
import logisim.comp.ComponentEvent;
import logisim.comp.ComponentListener;
import logisim.data.Attribute;
import logisim.data.AttributeEvent;
import logisim.data.AttributeListener;
import logisim.instance.Instance;
import logisim.instance.StdAttr;
import logisim.std.wiring.Pin;

public class CircuitPins {
	private class MyComponentListener implements ComponentListener, AttributeListener {
		public void endChanged(ComponentEvent e) {
			appearanceManager.updatePorts();
		}

		public void componentInvalidated(ComponentEvent e) {
		}

		public void attributeListChanged(AttributeEvent e) {
		}

		public void attributeValueChanged(AttributeEvent e) {
			Attribute<?> attr = e.getAttribute();
			if (attr == StdAttr.FACING || attr == StdAttr.LABEL || attr == Pin.ATTR_TYPE)
				appearanceManager.updatePorts();
		}
	}

	private PortManager appearanceManager;
	private MyComponentListener myComponentListener;
	private Set<Instance> pins;

	CircuitPins(PortManager appearanceManager) {
		this.appearanceManager = appearanceManager;
		myComponentListener = new MyComponentListener();
		pins = new HashSet<>();
	}

	public void transactionCompleted(ReplacementMap repl) {
		// determine the changes
		Set<Instance> adds = new HashSet<>();
		Set<Instance> removes = new HashSet<>();
		Map<Instance, Instance> replaces = new HashMap<>();
		for (Component comp : repl.getAdditions())
			if (comp.getFactory() instanceof Pin) {
				Instance in = Instance.getInstanceFor(comp);
				boolean added = pins.add(in);
				if (added) {
					comp.addComponentListener(myComponentListener);
					in.getAttributeSet().addAttributeListener(myComponentListener);
					adds.add(in);
				}
			}
		for (Component comp : repl.getRemovals())
			if (comp.getFactory() instanceof Pin) {
				Instance in = Instance.getInstanceFor(comp);
				boolean removed = pins.remove(in);
				if (removed) {
					comp.removeComponentListener(myComponentListener);
					in.getAttributeSet().removeAttributeListener(myComponentListener);
					Collection<Component> rs = repl.getComponentsReplacing(comp);
					if (rs.isEmpty())
						removes.add(in);
					else {
						Component r = rs.iterator().next();
						Instance rin = Instance.getInstanceFor(r);
						adds.remove(rin);
						replaces.put(in, rin);
					}
				}
			}

		appearanceManager.updatePorts(adds, removes, replaces, getPins());
	}

	public Collection<Instance> getPins() {
		return new ArrayList<>(pins);
	}
}
